java-client/src/main/java/org/opensearch/client/util/PathEncoder.java

```diff
@@ -8,59 +8,37 @@
 
 package org.opensearch.client.util;
 
-/*
- * SPDX-License-Identifier: Apache-2.0
- *
- * The OpenSearch Contributors require contributions made to
- * this file be licensed under the Apache-2.0 license or a
- * compatible open source license.
- */
-
-import java.lang.invoke.MethodHandle;
-import java.lang.invoke.MethodHandles;
-import java.lang.invoke.MethodType;
-import java.nio.charset.Charset;
-import java.nio.charset.StandardCharsets;
-import java.util.Collections;
-
 public class PathEncoder {
-    private final static String HTTP_CLIENT4_UTILS_CLASS = "org.apache.http.client.utils.URLEncodedUtils";
-    private final static String HTTP_CLIENT5_UTILS_CLASS = "org.apache.hc.core5.net.URLEncodedUtils";
-    private final static MethodHandle FORMAT_SEGMENTS_MH;
-
-    static {
-        Class<?> clazz = null;
-        try {
-            // Try Apache HttpClient5 first since this is a default one
-            clazz = Class.forName(HTTP_CLIENT5_UTILS_CLASS);
-        } catch (final ClassNotFoundException ex) {
-            try {
-                // Fallback to Apache HttpClient4
-                clazz = Class.forName(HTTP_CLIENT4_UTILS_CLASS);
-            } catch (final ClassNotFoundException ex1) {
-                clazz = null;
-            }
+    /**
+     * Percent encoding codec that matches Apache HTTP Client 4's path segment encoding.
+     */
+    @Deprecated
+    public static final PercentCodec APACHE_HTTP_CLIENT_4_EQUIV_CODEC = PercentCodec.RFC3986_PATHSAFE;
+    /**
+     * Percent encoding codec that matches Apache HTTP Client 5's path segment encoding.
+     */
+    public static final PercentCodec APACHE_HTTP_CLIENT_5_EQUIV_CODEC = PercentCodec.RFC3986_UNRESERVED;
+
+    public static final PercentCodec DEFAULT_CODEC = APACHE_HTTP_CLIENT_5_EQUIV_CODEC;
+
+    private static PercentCodec codec;
+
+    public static PercentCodec getCodec() {
+        if (codec == null) {
+            codec = DEFAULT_CODEC;
         }
+        return codec;
+    }
 
-        if (clazz == null) {
-            throw new IllegalStateException(
-                "Either '" + HTTP_CLIENT5_UTILS_CLASS + "' or '" + HTTP_CLIENT4_UTILS_CLASS + "' is required by not found on classpath"
-            );
-        }
+    public static void setCodec(PercentCodec codec) {
```

Comment:
Don't we want to let users call it for backwards compat?

Reply:
> Don't we want to let users call it for backwards compat?

May be not directly? As part of client transport builder?

Reply:
@reta The problem with calling it as part of the transport builder is you're then hiding the static/global nature of it behind an instance creation. It's perfectly reasonable that in my app I construct two different clients using different transports and could unwittingly override the setting. Which ideally it would be completely tied to the transport instance, but that's not backportable to 2.x.

Reply:
> @reta The problem with calling it as part of the transport builder is you're then hiding the static/global nature of it behind an instance creation.

Thanks @Xtansia , I understand that. Consider the alternative - asking users to call this method directly. Personally, I don't think we are giving the right tool out - I am having hard time even to explain in docuentation what are the consequences of letting to change the path encoder arbitrarily, from anywhere, anytime.

> It's perfectly reasonable that in my app I construct two different clients using different transports and could unwittingly override the setting. 

We have all the instruments to prevent that from happening: fe, use set-once semantics.

Reply:
I think the safest/cleanest/clearest way of having this as a global setting is using a system property at least for being backportable to 2.x. 

3.x could be refactored to attach it to the transport configuration and be per-instance.

Reply:
> If we implement this then we don't need a global public setting, do we? It's a convenience that causes a lot of side-effects, maybe we better force users to explicitly set it for every client they create?

That would be ideal  but the problem comes from usage of the path encoder: it is referenced statically from (literally) every single request class: `SimpleEndpoint.pathEncode(...);`. That's the showstopper that @Xtansia tries to solve with static `PathEncoder` configuration .

So if we do follow the builder pattern

```
builder = ApacheHttpClient5TransportBuilder.builder(hosts)
  .pathEncoder(PercentCodec.RFC3986_UNRESERVED)
  .build()
```

The `pathEncoder` would set the static path encoder behind the scene (and I think this is 100% fine). I hardly imagine the case when users would want clients with same JVM / app but different path encoders.

PS: I don't think we need decoder for path but that's just a comment.

Reply:
Ok, I understand. I am strongly opposed to an instance `.pathEncoder` that sets a global default. That's definitely not expected! We should either fix the static calls or I'm fine with a global setting.

Reply:
> That's definitely not expected! We should either fix the static calls or I'm fine with a global setting.

My mental model breaks static calls (sorry), but global setting (system property)  is looking acceptable, thanks @Xtansia

Reply:
Ok, I'm good with this.